include/mlir/SPIRV/SPIRVTypes.h

```diff
@@ -52,9 +53,28 @@ enum Kind {
 };
 }
 
+// Provides a common base class between SPIR-V ArrayType, SPIR-V StructType.
+class CompositeType : public Type {
+public:
+  using ImplType = detail::CompositeTypeStorage;
+  using Type::Type;
+
+  static bool kindof(unsigned kind) {
+    return (kind == TypeKind::Array || kind == TypeKind::Struct);
+  }
+
+  // Consumes an array of 32-bit indices.
+  // Returns a pair with the type selected by the last provided index on
+  // success, in other case returns a nullptr with an error message.
+  std::pair<Type, StringRef>
```

Comment:
I see we have `getMemberType` in this class too. Then it seems to me we don't need this method anymore. This way we don't need to handle the error cases inside the type library (and force the same error message to all callers). At call sites we can do something like

```cpp
for (int64_t index : indices) {
  if (auto cType = type.dyn_cast<CompositeType>()) {
    Type elemType = cType.getMemberType(index);
    if (!elemType) return ...emitError("cannot index into type ") << type << " with index " << index;
    type = elemType;
  } else {
    return ...emitError("cannot index into non-composite type ") << type;
  }
}
```
?

Reply:
Yea, seems more cleaner. Updated.